### PR TITLE
Open source d16n documentation

### PR DESCRIPTION
**Pull request by cakemanny**
>[!TIP]
> For reviewers: The spec looks best when rendered with asciidoctor but viewing on branch, i.e. https://github.com/bettermarks/d16n/blob/open-source-d16n-documentation/spec/d16n-v1_0.adoc , at least includes the images correctly. Which the PR view does not.

**README**

The general structure starts like https://github.com/bettermarks/opal/blob/main/README.md but stops once we get to instructions related to code, since there is no code here.

The table of implementers is inspired by the [RISC-V architecture ID document](https://github.com/riscv/riscv-isa-manual/blob/20240411/marchid.md?plain=1) where they have a table of implementers and their contact details.
Having knowledge that certain firms implement this spec might convince others to implement.

**spec/**

- Uses ascii doc instead of markdown as most advanced features of markdown are custom to each vendor. When writing a spec, it feels better to use a markup language which itself has a spec.

- Inspired by the [OpenID Connect spec](https://openid.net/specs/openid-connect-core-1_0.html), the [LTI spec](https://www.imsglobal.org/spec/lti/v1p3/) and RFCs in general.

There are still quite a number of todos and places for improvement
The overview diagram would be better if it gave a clear impression of the personal data not "escaping".
Most others are expressed in the comments within, right now

~I would mostly ask for a review of the content now and would come back and add a workflow to build and publish the rendered spec to gh-pages in a follow-up PR.~ _also done now_

Ideas for improvements/expansion:
- Clemens: add notes about CSP headers in the Implementation section.
- [x] Olli: add a license

---

Relates to https://bettermarks.atlassian.net/browse/CX-979

### CODE REVIEW

**Comment by cakemanny** [comment]
I feel like it might still be possible to improve the overview diagram with pikchr. I got some inspiration from their branching diagram [guide](https://pikchr.org/home/doc/trunk/doc/teardown01.md) to add some background boxes.


```
fill = white
linewid *= 0.5

down
RP: box "RP Server" radius 4px
arrow 0.75 from RP.s "2. access" aligned above "token" aligned below
RPC: box same "RP Client"

right

IDP: box same "Identity" "Provider " "(IDP)" at 2.0 right of RP.e
line from RP.e to IDP chop "1. authorization" above <->

arrow from RPC.e to IDP.sw <-> "3. retrieve Clear Names" aligned above

box \
  width RP.wid+linewid \
  height dist(RP.n,RPC.s)+linewid \
  behind RP fill lightyellow thin color grey \
  at dist(RP,RPC)/2 heading 180 from RP

box same \
  fill lightblue \
  at dist(RP,IDP) right of last box
```
overview-alt

>[!NOTE]
Part of the ugliness comes from github et al displaying this in a serif font. in the [pikchrshow](https://pikchr.org/home/pikchrshow) editor it shows it with san-serif...

File: README.md [hunk 1]
```diff
@@ -1 +1,109 @@
...
+Do we want to have the whole specification in this README? or should it be a
+separate doc? e.g. it could be a generated html page served on github pages...
+🤷
+
+-->
+
+## Implementers
+
+The following providers and apps have implemented this.
+Implementers can open a pull request against this repository to add their
```
Comment (JaschaSteudler) [inline README.md:97]: We should protect the main branch in this project if and introduce some PR rules?

**Review by JaschaSteudler — COMMENTED** [review]
Overall really cool. How do we proceed here with the open TODOs? I we want to give this to real IDPs there is some work to do

**Comment by cakemanny** [comment]
Another kind of idea / direction for an overview image. Maybe for the readme... or could go in the spec. (just brainstorming).

d16n-overview drawio

[Link to Edit in draw.io](https://viewer.diagrams.net/?tags=%7B%7D&highlight=0000ff&edit=_blank&layers=1&nav=1&title=d16n-overview.drawio#R7Vvbcps6FP0aZtKHepDEzY%2FxJT3JnPakSdPLU0fGMqbByAU5tfv1RwIJgzCN7cDYnYkz46CNbuyltbQlYQMNF%2Bt3CV7O39MpiQxoTtcGGhkQutDj38KwyQ2WZeeGIAmnuQlsDffhbyKNprSuwilJKxkZpRELl1WjT%2BOY%2BKxiw0lCf1WzzWhUbXWJA1Iz3Ps4qlu%2FhFM2z622h7b2f0gYzGXLFpA3FljllQ%2BSzvGU%2FiqZ0NhAw4RSll8t1kMSCdcpt%2BTlrhruFv1KSMz2KfBxQt2bm493Tx6w4%2BF97N48PL6FeS1POFrJ5zWgE%2FH6BjPKqxVujWiS3XF%2BrkRXBwZELsEOMcsmJxD%2Fv5AJLzIQTieJIfpzxb%2Ff00nIvQnNy%2BVS1c7v5A3kBaWH2EZ5HafLHMxZuCb8AQZztoh4EvDLJQ1jlqFqDwx7xC04CoOYG3zuCt4wGoSLDNXsIeSQAnBrH4WLgPcgCnl3r%2FDvVUKE3wISkwRzl13JJ%2BilT0HRuSeSMLJu9D0oEOVEIHRBWLLhWVQBUw4CSQLgySp%2BbYeUI3kxL40mW%2BKD5SAOipq3QPMLifUBuFut4v5wP747Z2S5gsQsZNxtVw8c17Q9YGEVV6TSZVytOq6u2RGuTqu4Xo%2FGHz5df%2FrGy93e%2FfeZJznM5sX16PbNXwL3tbz8%2Fo5ynGMc%2B6Q98C2d1a69F%2FrK1jr6Xqvoc1J%2Fvh6OK%2BCfL%2Bhb9b4nCYf1%2BxVOFi0S3dGYvh%2FRO4NahUcHYu1knzrWd%2BTniqQcFxFdCTAJUdVNEpUpClNRMZ3xr5XQ0X3HA5nykEomYxoLbGnC5jSgMY7%2BpXQpx8YPwthGIo5XvCOVkZPQVTwVw2lk8hRZh%2ByruO5ZniXT33j6rdkzbVcaRutS7tGmlLglSciREOMtt8Ucla%2ByvKsM37IGbOQow7bCLLUpp%2FQqZ2EUDQsQ0MwWf9yesoQ%2BktIdCYq4gxN2KYLWraPy3Cr0FDSY4nSeuQHIxC1mvNU4s0ATFAgIt%2F95uHOU6CrxyfORAu9ZQNhzkWQjfbhXbajkcVPpQYlBnhzUCYkwC5%2Bqnd9FIdnerdCUUrSFYM8G%2Fe3Hq7QLHdRzeURWfDSO5i6RdZYDaq2ZYhZXkgC0WT33Wa0iDjDelLJJRWx8HLvaDDT7mnrkFWqFVe10NksJM3S9KSA6XoIUMK1LkK472PdJmkvTI4lPKjpllbCrKuGasA2V8Hzi%2B7tUYuLZlm3uVIStFrquo2mh1T9YC1tUD7inenh%2FVA%2Bzh2wTVVjwdkek3ZZ82HpYp%2FF6b4GwtIqcbgRC77DU2dMLxHG7C00Ccbni7E3ClAgpoHFdJyZkjqOZilAesj2I02nF4cRWNAY9E4ESjYW2oINZvEMmlHpxPpn9snplQgGOka8WtcLbUyucZ7TCdtTuitIKqzOtQPoSUKf4vlpRX0t2oxUWPFOtOG5HqkkrrtN0RYqQQVxcFFEEjqcZ8LOEpPM3vbOIJ4BRiSVg3z2GjC3HEsjTYwnvYBVqUR%2BclmIJE7mwqg%2FA60ofLD2WsI7Vh%2F4zFbWlD5Zbacfaa7FRbHNXyqJqD7sTD3TctmfzSoStkkw0zmjlIWmZ0dA1ysEBtA9mZRtCIVxQyumOL51xFxHBn%2FYUEFDzRkFlpysqQ22GPnamB9oMDLyOVgVaO5ZzHjM9Om6Xeo%2Bdy2FEMA%2F5zQ94QS7SN%2FUFQrYuoGyeHU6ezQ4meCmbq3uXXjWUQACdKpSQA1nVfL%2FEfhgHanmS02X3zR%2BrxVJzd4uysu%2BmxHMLDdNzKgyD9eOftrRHm1lB%2F0jxQZ6%2BmfhS8ak31b2EqNin7fle14sGQTmz%2Bd9x4fGSAXom0Dcywck2Mlsk%2Bb6rBfgcyfv9aoihdrU6YLlOTnjsxqMWqsAXn0ychOX1qB7WuMbImmm8IWn4G0%2ByDGKUCWal8hTZ2HHaLI6FQx9Hl%2FLGIpxOReEdJ9VVShYvdIlEEOE0VTceCfPnihbVMHmgBn92uC2fQx2kvOgVI32lCeqbXurQunxCjbSh0d4rRm4NPvQKXxN8SFvfe6dGrx6lW6%2FoNaGnb%2FOo1z5Ohl6%2Fhp79il4Tev3zAs%2BuR7fOK3hN4OmxTnECczL46q9FuK%2FwNc57oCHmPRl8sAYf%2BOvhK94wa599VfTku0kdoMeT299K5EuM7e9N0Ph%2F)

<details><summary><code>d16-overview.drawio</code></summary>
<p>

`<mxfile host="app.diagrams.net" modified="2024-11-14T20:04:28.715Z" agent="5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36" etag="o68mfTZ08mqtoR3fYCzc" version="20.3.7" type="device"><diagram id="NorhpUGhAtCH0RtukjRp" name="Page-1">7Vvbcps6FP0aZtKHepDEzY/xJT3JnPakSdPLU0fGMqbByAU5tfv1RwIJgzCN7cDYnYkz46CNbuyltbQlYQMNF+t3CV7O39MpiQxoTtcGGhkQutDj38KwyQ2WZeeGIAmnuQlsDffhbyKNprSuwilJKxkZpRELl1WjT+OY+Kxiw0lCf1WzzWhUbXWJA1Iz3Ps4qlu/hFM2z622h7b2f0gYzGXLFpA3FljllQ+SzvGU/iqZ0NhAw4RSll8t1kMSCdcpt+TlrhruFv1KSMz2KfBxQt2bm493Tx6w4+F97N48PL6FeS1POFrJ5zWgE/H6BjPKqxVujWiS3XF+rkRXBwZELsEOMcsmJxD/v5AJLzIQTieJIfpzxb/f00nIvQnNy+VS1c7v5A3kBaWH2EZ5HafLHMxZuCb8AQZztoh4EvDLJQ1jlqFqDwx7xC04CoOYG3zuCt4wGoSLDNXsIeSQAnBrH4WLgPcgCnl3r/DvVUKE3wISkwRzl13JJ+ilT0HRuSeSMLJu9D0oEOVEIHRBWLLhWVQBUw4CSQLgySp+bYeUI3kxL40mW+KD5SAOipq3QPMLifUBuFut4v5wP747Z2S5gsQsZNxtVw8c17Q9YGEVV6TSZVytOq6u2RGuTqu4Xo/GHz5df/rGy93e/feZJznM5sX16PbNXwL3tbz8/o5ynGMc+6Q98C2d1a69F/rK1jr6Xqvoc1J/vh6OK+CfL+hb9b4nCYf1+xVOFi0S3dGYvh/RO4NahUcHYu1knzrWd+TniqQcFxFdCTAJUdVNEpUpClNRMZ3xr5XQ0X3HA5nykEomYxoLbGnC5jSgMY7+pXQpx8YPwthGIo5XvCOVkZPQVTwVw2lk8hRZh+yruO5ZniXT33j6rdkzbVcaRutS7tGmlLglSciREOMtt8Ucla+yvKsM37IGbOQow7bCLLUpp/QqZ2EUDQsQ0MwWf9yesoQ+ktIdCYq4gxN2KYLWraPy3Cr0FDSY4nSeuQHIxC1mvNU4s0ATFAgIt/95uHOU6CrxyfORAu9ZQNhzkWQjfbhXbajkcVPpQYlBnhzUCYkwC5+qnd9FIdnerdCUUrSFYM8G/e3Hq7QLHdRzeURWfDSO5i6RdZYDaq2ZYhZXkgC0WT33Wa0iDjDelLJJRWx8HLvaDDT7mnrkFWqFVe10NksJM3S9KSA6XoIUMK1LkK472PdJmkvTI4lPKjpllbCrKuGasA2V8Hzi+7tUYuLZlm3uVIStFrquo2mh1T9YC1tUD7inenh/VA+zh2wTVVjwdkek3ZZ82HpYp/F6b4GwtIqcbgRC77DU2dMLxHG7C00Ccbni7E3ClAgpoHFdJyZkjqOZilAesj2I02nF4cRWNAY9E4ESjYW2oINZvEMmlHpxPpn9snplQgGOka8WtcLbUyucZ7TCdtTuitIKqzOtQPoSUKf4vlpRX0t2oxUWPFOtOG5HqkkrrtN0RYqQQVxcFFEEjqcZ8LOEpPM3vbOIJ4BRiSVg3z2GjC3HEsjTYwnvYBVqUR+clmIJE7mwqg/A60ofLD2WsI7Vh/4zFbWlD5Zbacfaa7FRbHNXyqJqD7sTD3TctmfzSoStkkw0zmjlIWmZ0dA1ysEBtA9mZRtCIVxQyumOL51xFxHBn/YUEFDzRkFlpysqQ22GPnamB9oMDLyOVgVaO5ZzHjM9Om6Xeo+dy2FEMA/5zQ94QS7SN/UFQrYuoGyeHU6ezQ4meCmbq3uXXjWUQACdKpSQA1nVfL/EfhgHanmS02X3zR+rxVJzd4uysu+mxHMLDdNzKgyD9eOftrRHm1lB/0jxQZ6+mfhS8ak31b2EqNin7fle14sGQTmz+d9x4fGSAXom0Dcywck2Mlsk+b6rBfgcyfv9aoihdrU6YLlOTnjsxqMWqsAXn0ychOX1qB7WuMbImmm8IWn4G0+yDGKUCWal8hTZ2HHaLI6FQx9Hl/LGIpxOReEdJ9VVShYvdIlEEOE0VTceCfPnihbVMHmgBn92uC2fQx2kvOgVI32lCeqbXurQunxCjbSh0d4rRm4NPvQKXxN8SFvfe6dGrx6lW6/oNaGnb/Oo1z5Ohl6/hp79il4Tev3zAs+uR7fOK3hN4OmxTnECczL46q9FuK/wNc57oCHmPRl8sAYf+OvhK94wa599VfTku0kdoMeT299K5EuM7e9N0Ph/</diagram></mxfile>`

</p>
</details> 

**Edit:**

This was also a draft I tried... 🤷  
d16n-fat-arrows drawio

_I think the key has become deformed... but oh well_

[Link to view/edit in draw.io](https://viewer.diagrams.net/?tags=%7B%7D&lightbox=1&highlight=0000ff&edit=_blank&layers=1&nav=1#R%3Cmxfile%3E%3Cdiagram%20name%3D%22Seite-1%22%20id%3D%22ZTEOLTrBGfIjOySsdjNL%22%3E3ZnLbuIwFIafhuWgOE5CWEJpy0hTCYlWlWaDTOImVhM7csxtnn5s4kBM3MtIgWlRF7GP7%2F%2FnYx%2BXHrzJt%2FccFekDi3HWc51424OTnusCxw3kR1l2lSXwncqQcBLrSkfDnPzBdUttXZEYl0ZFwVgmSGEaI0YpjoRhQ5yzjVnthWXmqAVKcMswj1DWtj6TWKSVNXQHR%2FsUkyStRwbBsCrJUV1Zr6RMUcw2DRO87cEbzpioUvn2BmdKvFqXqt3dG6WHiXFMxWcaDIF7d%2FeIBsXv0eOuDAdP8H76o57cGmUrvWI9W7GrJdikROB5gSKV30jMPThORZ7JHJDJGJUpjnVG94a5wNs35wkOq5fbBrMcC76TVeoGNXq9Yzyd3RzlBwNtSxvSw1AbkUaeHLo%2BqiITWph%2FEAl8QqSGIAfQTlMdlUEZSahMR1IKzKVBCUXkZhvpgpzEsepR9VGonvNtovyqX21lt%2FqqbpU4Tj9QSaWbp7p%2FyUgx1ZOgTESpHpYzgQRhagDP7waSG3p9YGACnlNbmqTcsE1qcDZQ4GNQnK1ofCBy6b3tm3vb9dubG9o2N%2FDOpdnwe%2B%2FtLvby8ARKYDlxHAuUs%2B1j93swiQmXV251sJRspbTqmg0Y%2BuYpY3EYq78452LjfcyG5PugYly%2B4qMYqCyq%2BOSFbBWg5jFTMELFfqL%2BuOdPbOReGBU6OgJSr7EeY0LyRC4jI0u1mDJCWH6f8XLKSkFo0i%2FXSUdOAkwQYZuDb%2BHgnwuD%2F%2FUx%2FIxlKyJ2i3smpaeIRrg7Hh48ObQG%2FxlI8PWBPJWYL2acrFG0644E9EwS0LXc6cEFSQw%2BJmEQ2KuM%2Be0aV2K%2Ff6c0Lx7B2Su%2BYRmTJCaUUQU3Q0uczVhJ9MXQunt%2BnVRYMiFYbrmcBCusxEmW1WP2XOg76s9yhe2p9%2BXSVhT3Y7wmEV6wsouj0DEjX9tZ6A0vCDz8%2Bq73gKgsjBf6TCS47NABB6YDAmhxQMsz5Gw8QDt%2Bm2FeMopU0wkSqMVHrl685VHveoCu0yD3Zoxne%2BuYr6EunCM8OQ0tz%2FfA9no%2FGwzYglFfzNI642xNYqnq1QLxXDNws11PlwXSjqDVxaw8Y39IXy8KcPLQhJaD6rIo2lH03Yrun3RSWLG7YhTw5F1pefNfFkU7fh4VhTTMMV9f8%2FkEPDOasgRTlwXRDp9HUYRLGTkGmdj%2F12Mtk4lKPkrtaW2XozWKrheY54Sm65zvhpfZ4%2B8j%2B7LGr0zw9i8%3D%3C%2Fdiagram%3E%3C%2Fmxfile%3E)

**Comment by cakemanny** [comment]
With Gerald, we tried to approach producing a diagram for the readme:
this is what we sketched in miro:

image
> [!WARNING]
> This image has been corrected. It should read **Authorization** and NOT **Authentication**

I think we would like to recreate this outside of Miro in order to have it be editable. Preferably in some diagram as code form.

**Review by JaschaSteudler — APPROVED** [review]

**Pull request merged**